File: appupdatelibrary/src/main/java/com/dcastalia/localappupdate/DownloadApk.java
package com.dcastalia.localappupdate;

import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Build;
import android.os.Environment;
import android.util.Log;
import android.widget.Toast;
import androidx.core.content.FileProvider;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;

/**
 * Created by piashsarker on 1/16/18.
 */

public class DownloadApk extends Activity{


    private static ProgressDialog bar;
    private static String TAG = "DownloadApk";
    private static Context context ;
    private static Activity activity ;
    private static String downloadUrl ;


    public DownloadApk(Context context){
        this.context = context ;
        this.activity = (Activity)context;
    }


    public   void startDownloadingApk(String url){
           downloadUrl = url ;
           if(downloadUrl!=null){
               new DownloadNewVersion().execute();
           }

    }




    private static  class DownloadNewVersion extends AsyncTask<String,Integer,Boolean> {

        @Override
        protected void onPreExecute() {
            super.onPreExecute();
            if(bar==null){
                bar = new ProgressDialog(context);
                bar.setCancelable(false);
                bar.setMessage("Downloading...");
                bar.setIndeterminate(true);
                bar.setCanceledOnTouchOutside(false);
                bar.show();
            }


        }

        protected void onProgressUpdate(Integer... progress) {
            super.onProgressUpdate(progress);


            bar.setIndeterminate(false);
            bar.setMax(100);
            bar.setProgress(progress[0]);
            String msg = "";
            if(progress[0]>99){

                msg="Finishing... ";

            }else {

                msg="Downloading... "+progress[0]+"%";
            }
            bar.setMessage(msg);

        }
        @Override
        protected void onPostExecute(Boolean result) {
            // TODO Auto-generated method stub
            super.onPostExecute(result);
            if(bar.isShowing() && bar!=null){
                bar.dismiss();
                bar=null;
            }


            if(result){

                Toast.makeText(context,"Update Done",
                        Toast.LENGTH_SHORT).show();

            }else{

                Toast.makeText(context,"Error: Try Again",
                        Toast.LENGTH_SHORT).show();

            }

        }

        @Override
        protected Boolean doInBackground(String... arg0) {
            Boolean flag = false;
            try {
                URL url = new URL(downloadUrl);
                HttpURLConnection c = (HttpURLConnection) url.openConnection();
                c.setRequestMethod("GET");
                c.connect();
                String PATH = Environment.getExternalStorageDirectory()+"/Download/";
                File file = new File(PATH);
                file.mkdirs();
                File outputFile = new File(file,"app-debug.apk");

                if(outputFile.exists()){
                    outputFile.delete();
                }

                FileOutputStream fos = new FileOutputStream(outputFile);
                InputStream is = c.getInputStream();

                float total_size = c.getContentLength();//size of apk

                byte[] buffer = new byte[1024];
                int len1 = 0;
                float per = 0;
                float downloaded=0;
                while ((len1 = is.read(buffer)) != -1) {
                    fos.write(buffer, 0, len1);
                    downloaded +=len1;
                    per = (downloaded * 100 / total_size);
                    publishProgress((int) per);
                }
                fos.close();
                is.close();
                OpenNewVersion(PATH);
                flag = true;
            } catch (MalformedURLException e) {
                Log.e(TAG, "Update Error: " + e.getMessage());
                flag = false;
            }catch (IOException ex){
                ex.printStackTrace();
            }
            return flag;

        }

    }

    private static void OpenNewVersion(String location) {
        Intent intent = new Intent(Intent.ACTION_VIEW);
        intent.setDataAndType(getUriFromFile(location),
                "application/vnd.android.package-archive");
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
        context.startActivity(intent);

    }

    private static Uri getUriFromFile(String location) {

        if(Build.VERSION.SDK_INT<24){
            return   Uri.fromFile(new File(location + "app-debug.apk"));
        }
        else{
            return FileProvider.getUriForFile(context,
                    context.getApplicationContext().getPackageName() + ".provider",
                    new File(location + "app-debug.apk"));
        }
    }







}

File: appupdatelibrary/src/main/java/com/dcastalia/localappupdate/DownloadApk.kt
package com.dcastalia.localappupdate

import android.app.ProgressDialog
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.AsyncTask
import android.os.Build
import android.os.Environment
import android.util.Log
import android.webkit.URLUtil
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.FileProvider
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL

/**
 * Created by piashsarker on 1/16/18, rewritten by Cyb3rKo on 12/21/20.
 */

class DownloadApk(var context: Context) : AppCompatActivity() {

    @JvmOverloads
    fun startDownloadingApk(url: String, fileName: String = "App Update") {
        if (URLUtil.isValidUrl(url)) {
            DownloadNewVersion(context, url, fileName).execute()
        }
    }

    @Suppress("DEPRECATION")
    private class DownloadNewVersion(
        val context: Context,
        val downloadUrl: String,
        val fileName: String
    ): AsyncTask<String, Int, Boolean>() {
        private lateinit var bar: ProgressDialog
        override fun onPreExecute() {
            super.onPreExecute()
            bar = ProgressDialog(context).apply {
                setCancelable(false)
                setMessage("Downloading...")
                isIndeterminate = true
                setCanceledOnTouchOutside(false)
                show()
            }
        }

        override fun onProgressUpdate(vararg values: Int?) {
            super.onProgressUpdate(*values)
            var msg = ""
            val progress = values[0]
            if (progress != null) {
                bar.progress = progress
                msg = if (progress > 99) "Finishing... " else "Downloading... $progress%"
            }

            bar.apply {
                isIndeterminate  = false
                max = 100
                setMessage(msg)
            }
        }

        override fun onPostExecute(result: Boolean?) {
            super.onPostExecute(result)
            bar.dismiss()
            if (result != null && result) {
                Toast.makeText(context, "Update Done Dona Done Done", Toast.LENGTH_SHORT).show()
            } else {
                Toast.makeText(context, "Update Failed. Please Update From Website", Toast.LENGTH_SHORT).show()
            }
        }

        override fun doInBackground(vararg p0: String?): Boolean {
            var flag = false

            try {
                val path = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS).toString() + "/"
                var outputFile = File("$path$fileName.apk")
                var repetition = 1
                while (outputFile.exists()) {
                    outputFile = File("$path$fileName ($repetition).apk")
                    repetition++
                }

                val directory = File(path)
                if (!directory.exists()) {
                    directory.mkdirs()
                }

                val url = URL(downloadUrl)
                val c = url.openConnection() as HttpURLConnection
                c.requestMethod = "GET"
                c.connect()

                val fos = FileOutputStream(outputFile)
                val inputStream = c.inputStream
                val totalSize = c.contentLength.toFloat() //size of apk

                val buffer = ByteArray(1024)
                var len1: Int
                var per: Float
                var downloaded = 0f
                while (inputStream.read(buffer).also { len1 = it } != -1) {
                    fos.write(buffer, 0, len1)
                    downloaded += len1
                    per = (downloaded * 100 / totalSize)
                    publishProgress(per.toInt())
                }
                fos.close()
                inputStream.close()
                openNewVersion(outputFile.path)
                flag = true
            } catch (e: MalformedURLException) {
                Log.e("DownloadApk", "Update Error: " + e.message)
                flag = false
            } catch (e: IOException) {
                e.printStackTrace()
            }

            return flag
        }

        private fun openNewVersion(location: String) {
            val intent = Intent(Intent.ACTION_VIEW)
            intent.setDataAndType(
                getUriFromFile(location),
                "application/vnd.android.package-archive"
            )
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
            context.startActivity(intent)
        }

        private fun getUriFromFile(filePath: String): Uri {
            return if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
                Uri.fromFile(File(filePath))
            } else {
                FileProvider.getUriForFile(
                    context,
                    context.packageName + ".provider",
                    File(filePath)
                )
            }
        }
    }
}